### ISSUE TITLE
[TF 2.0] Inconvenient graph visualization

### ISSUE THREAD

**Issue by llan-ml** [tensorflow/tensorflow#29215]
**System information**
- TensorFlow version (use command below): 2.0.0-alpha0
- Python version: 3.6.5

When visualizing the graph with tensorboard, there are many nodes, which maybe generated by auotgraph, to make it quite inconvenient to debug with the graph.

To reproduce:
<pre>
import numpy as np
import tensorflow as tf

class Model(tf.keras.Model):
    def __init__(self):
        super(Model, self).__init__()
        self.dense1 = tf.keras.layers.Dense(100)

    def call(self, inputs):
        outputs = self.dense1(inputs)
        return outputs

model = Model()
optimizer = tf.keras.optimizers.Adam()
writer = tf.summary.create_file_writer("./logdir")

@tf.function
def train(data):
    with tf.name_scope("xxx"):
        with tf.GradientTape() as tape:
            y = model(data)
            loss = tf.reduce_mean(tf.square(y))
        grads = tape.gradient(loss, model.trainable_weights)
        optimizer.apply_gradients(zip(grads, model.trainable_weights))

x = np.random.rand(10, 100).astype(np.float32)
# y = model(x)
# train.python_function(x)
tf.summary.trace_on()
train(x)
with writer.as_default():
    tf.summary.trace_export("graph", step=0)
    tf.summary.trace_off()
    writer.flush()
</pre>

The graph visualization contains many `ReadVariableOp` and `AssignVariableOp` outside, as follows:
![png](https://user-images.githubusercontent.com/22030149/59081332-196c5680-8920-11e9-9b76-bfa482ccf2b2.png)

If we unocmment `y = model(x)`, the number of outside ops decrease, as follows:
![png (1)](https://user-images.githubusercontent.com/22030149/59081425-9bf51600-8920-11e9-8dfb-4c3b813f9f5c.png)

If we uncomment `train.python_function(x)`, all `ReadVariableOp` and `AssignVariableOp` decrease:
![png (2)](https://user-images.githubusercontent.com/22030149/59081460-c6df6a00-8920-11e9-8448-6e368ef5d573.png)

I think this is quite related to #29442 . The keras layers add variables during the first call. If the first call is inside `tf.function`, these variables are created in a graph and some unexpected behaviors happen as reported in #29442 .

Uncommenting `y = model(x)` makes sure that the model variables are created in eager mode, and uncommenting `train.python_function(x)` makes sure that optimizer-related variables are also created in eager mode.

**Comment by gadagashwini** [comment]
@llan-ml Code snippet looks incomplete could you provide the complete code to reproduce the issue. Thanks!

**Comment by llan-ml** [comment]
@gadagashwini I updated the code snippet above.

**Comment by mdanatg** [comment]
@llan-ml Autograph doesn't affect name scopes in the default configuration, so it should not affect the graph display. Can you re-run the snippet setting `@tf.function(autograph=False)` to confirm?

@fchollet Any thoughts about which ops are created outside the `name_scope` inside `train`?

**Comment by llan-ml** [comment]
@mdanatg Aftering setting `autograph=False`, there is no difference.

As shown in the above picture, the `ReadVariableOp` and `AssignVariableOp` are outside.

**Comment by llan-ml** [comment]
@mdanatg I updated the issue and it shows more findings now.

**Comment by mdanatg** [comment]
Thanks for double-checking. I'm not sure what creates the ops outside the name scope - I suspect it's either the optimizer or the model tracer.

**Comment by bionicles** [comment]
how do you add a keras model to tensorflow graph when you're using GradientTape? No reference available on this ... docs only describe the callbacks with model.fit, seems impossible to add a keras model to tensorboard if using custom training loop

**Comment by llan-ml** [comment]
@bionicles The decorator `tf.function` will automatically transform the code in a function into the graph-style code, and `tf.summary.trace_on` will trace graph execution.

**Comment by bionicles** [comment]
@llan-ml thank you

**Comment by tomerk** [comment]
It's not clear to me that this is a Keras issue as opposed to a more general issue with how Tensorboard visualizes variable assignment operators when variables are captured from tf.functions vs. created in a run of the tf.function.

That said, we've made a lot of changes to Keras's execution path recently, so if you try the new nightly you *might* get a less noisy visualization. Meanwhile, re-assigning the more-general question of how to visualize assignment operators to the tensorboard team for triaging.

**Comment by manivaradarajan** [comment]
Reassigning to davidsoergel to prioritize for the Graph plugin.

**Comment by jvishnuvardhan** [comment]
I ran @llan-ml code in recently released `TF2.0` and see simpler graph than the one with `TF2.0.alpha0`. [Here](https://colab.sandbox.google.com/gist/jvishnuvardhan/129a7c6950a0c001da57319aa0adaad1/tf_29215.ipynb) is the gist with `TF2.0`. Thanks!

**Comment by llan-ml** [comment]
@jvishnuvardhan Thanks for your effort!

For now, there are still some extra internal ops existing like
![image](https://user-images.githubusercontent.com/22030149/66018682-9d230200-e512-11e9-9c3e-25f2415bde6f.png)

It would be better if either we can batch these extra ops (except for `data`) within a predefined or customized scope or they are removed from the main graph by default.

**Comment by 13ecd1** [comment]
> @bionicles The decorator `tf.function` will automatically transform the code in a function into the graph-style code, and `tf.summary.trace_on` will trace graph execution.

When I do this and open tensorboard using tensorboard --logdir logdirlocation only a scalars tab opens up. Do you happen to know why the graphs tab does not appear? I have tf.summary.trace_on defined after I instantiated my models.

**Comment by hassanhub** [comment]
I have the same problem with graph visualization.
I used to rely on TB graph visualization in TF 1.X for sanity checks and it was extremely useful.
Now, it's frustrating to normally keep track of the I/O flow of each block.
There are plenty of nodes with this naming pattern: **_XXXX_cast_readvariableop_resource_**
Putting a custom model in Keras also doesn't work for visualizing a conceptual graph.

**Comment by llan-ml** [comment]
For those who still suffer from this issue, I found that we can first convert the custom model into a functional model (i.e., initialize the model [in a graph manner](https://github.com/tensorflow/tensorflow/blob/25fba035f3e453d94490932096282c7b0624bbb3/tensorflow/python/keras/engine/network.py#L173)) and then visualize the graph with `tf.keras.callbacks.Tensorboard`, which usually results in a much better graph visualization than that generated as in this [link](https://www.tensorflow.org/tensorboard/graphs#graphs_of_tffunctions).

**Comment by llan-ml** [comment]
I believe that this issue has been resolved with the new public summary operation [`tf.summary.graph`](https://www.tensorflow.org/api_docs/python/tf/summary/graph).
